hazelcast/src/main/java/com/hazelcast/map/impl/operation/EntryOffloadableOperation.java

```diff
@@ -0,0 +1,298 @@
+    private transient SerializationService ss;
+    private transient long begin;
+
+    public EntryOffloadableOperation() {
+    }
+
+    public EntryOffloadableOperation(String name, Data dataKey, EntryProcessor entryProcessor) {
+        super(name, dataKey);
+        this.entryProcessor = entryProcessor;
+    }
+
+    @Override
+    public void innerBeforeRun() throws Exception {
+        super.innerBeforeRun();
+
+        this.ops = (OperationServiceImpl) getNodeEngine().getOperationService();
+        this.ss = getNodeEngine().getSerializationService();
+        this.begin = getNow();
+        this.readOnly = entryProcessor instanceof ReadOnly;
+
+        final SerializationService serializationService = getNodeEngine().getSerializationService();
+        final ManagedContext managedContext = serializationService.getManagedContext();
+        managedContext.initialize(entryProcessor);
+    }
+
+    @Override
+    public void run() {
+        if (!(entryProcessor instanceof Offloadable)) {
+            throw new HazelcastException("EntryProcessor is expected to implement Offloadable for this operation");
+        }
+        if (readOnly && entryProcessor.getBackupProcessor() != null) {
+            throw new HazelcastException("EntryProcessor.getBackupProcessor() should return null if ReadOnly implemented");
+        }
+
+        boolean shouldCloneForOffloading = OBJECT.equals(mapContainer.getMapConfig().getInMemoryFormat());
+        Object value = recordStore.get(dataKey, false);
+        value = shouldCloneForOffloading ? ss.toData(value) : value;
+
+        String executorName = ((Offloadable) entryProcessor).getExecutorName();
+        executorName = executorName.equals(Offloadable.DEFAULT_ASYNC_EXECUTOR) ? ASYNC_EXECUTOR : executorName;
+
+        if (readOnly) {
+            runOffloadedReadOnlyEntryProcessor(value, executorName);
+        } else {
+            runOffloadedModifyingEntryProcessor(value, executorName);
+        }
+    }
+
+    private void runOffloadedReadOnlyEntryProcessor(final Object previousValue, String executorName) {
+        // execute for key
+        getNodeEngine().getExecutionService().execute(executorName, new Runnable() {
+            @Override
+            public void run() {
+                try {
+                    final Map.Entry entry = createMapEntry(dataKey, previousValue);
+                    final Data result = process(entry);
+                    if (!noOp(entry, previousValue)) {
+                        throw new HazelcastException("It is not allowed to modify an entry in a ReadOnly EntryProcessor");
+                    }
+                    getOperationResponseHandler().sendResponse(EntryOffloadableOperation.this, result);
+                } catch (Throwable t) {
+                    getOperationResponseHandler().sendResponse(EntryOffloadableOperation.this, t);
+                } finally {
+                    ops.onCompletionAsyncOperation(EntryOffloadableOperation.this);
```

Comment:
I think you meant that in the read-only path it was not started but completed? I fixed that, thx.